File: php/test/UpdateResultEntityTest.php
<?php
declare(strict_types=1);

// UpdateResult entity test

require_once __DIR__ . '/../bluefinshieldconexmgmt_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class UpdateResultEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = BluefinShieldconexMgmtSDK::test(null, null);
        $ent = $testsdk->UpdateResult(null);
        $this->assertNotNull($ent);
    }

    // Feature #4: the entity stream(action, ...) method runs the op pipeline
    // and yields result items. With the streaming feature active it yields the
    // feature's incremental output; otherwise it falls back to the materialised
    // list so stream always yields.
    public function test_stream(): void
    {
        $seed = [
            "entity" => [
                "update_result" => [
                    "s1" => ["id" => "s1"],
                    "s2" => ["id" => "s2"],
                    "s3" => ["id" => "s3"],
                ],
            ],
        ];

        // Fallback: streaming inactive -> yields the materialised list items.
        $base = BluefinShieldconexMgmtSDK::test($seed, null);
        $seen = iterator_to_array($base->UpdateResult(null)->stream("list", null, null), false);
        $this->assertCount(3, $seen);

        // Inbound: streaming active -> yields each item from the feature.
        $cfg = BluefinShieldconexMgmtConfig::make_config();
        if (isset($cfg["feature"]) && is_array($cfg["feature"]) && isset($cfg["feature"]["streaming"])) {
            $sdk = BluefinShieldconexMgmtSDK::test($seed, ["feature" => ["streaming" => ["active" => true]]]);
            $got = [];
            foreach ($sdk->UpdateResult(null)->stream("list", null, null) as $item) {
                if (is_array($item) && array_is_list($item)) {
                    foreach ($item as $sub) {
                        $got[] = $sub;
                    }
                } else {
                    $got[] = $item;
                }
            }
            $this->assertCount(3, $got);
        }
    }

    public function test_basic_flow(): void
    {
        $setup = update_result_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create", "list", "update"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "update_result." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set BLUEFINSHIELDCONEXMGMT_TEST_UPDATE_RESULT_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $update_result_ref01_ent = $client->UpdateResult(null);
        $update_result_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.update_result"), "update_result_ref01"));

        $update_result_ref01_data_result = $update_result_ref01_ent->create($update_result_ref01_data, null);
        $update_result_ref01_data = Helpers::to_map($update_result_ref01_data_result);
        $this->assertNotNull($update_result_ref01_data);
        $this->assertNotNull($update_result_ref01_data["id"]);

        // LIST
        $update_result_ref01_match = [];

        $update_result_ref01_list_result = $update_result_ref01_ent->list($update_result_ref01_match, null);
        $this->assertIsArray($update_result_ref01_list_result);

        $found_item = sdk_select(
            Runner::entity_list_to_data($update_result_ref01_list_result),
            ["id" => $update_result_ref01_data["id"]]);
        $this->assertNotEmpty($found_item);

        // UPDATE
        $update_result_ref01_data_up0_up = [
            "id" => $update_result_ref01_data["id"],
        ];

        $update_result_ref01_markdef_up0_name = "billing_id";
        $update_result_ref01_markdef_up0_value = "Mark01-update_result_ref01_" . $setup["now"];
        $update_result_ref01_data_up0_up[$update_result_ref01_markdef_up0_name] = $update_result_ref01_markdef_up0_value;

        $update_result_ref01_resdata_up0_result = $update_result_ref01_ent->update($update_result_ref01_data_up0_up, null);
        $update_result_ref01_resdata_up0 = Helpers::to_map($update_result_ref01_resdata_up0_result);
        $this->assertNotNull($update_result_ref01_resdata_up0);
        $this->assertEquals($update_result_ref01_resdata_up0["id"], $update_result_ref01_data_up0_up["id"]);
        $this->assertEquals($update_result_ref01_resdata_up0[$update_result_ref01_markdef_up0_name], $update_result_ref01_markdef_up0_value);

    }
}

function update_result_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/update_result/UpdateResultTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = BluefinShieldconexMgmtSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["update_result01", "update_result02", "update_result03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("BLUEFINSHIELDCONEXMGMT_TEST_UPDATE_RESULT_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "BLUEFINSHIELDCONEXMGMT_TEST_UPDATE_RESULT_ENTID" => $idmap,
        "BLUEFINSHIELDCONEXMGMT_TEST_LIVE" => "FALSE",
        "BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN" => "FALSE",
        "BLUEFINSHIELDCONEXMGMT_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["BLUEFINSHIELDCONEXMGMT_TEST_UPDATE_RESULT_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["BLUEFINSHIELDCONEXMGMT_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["BLUEFINSHIELDCONEXMGMT_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new BluefinShieldconexMgmtSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["BLUEFINSHIELDCONEXMGMT_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}

File: cpp/test/update_result_entity_test.cpp
// Generated basic-flow test for the update_result entity (model-driven,
// unit mode; mirrors the rust/go TestEntity generator).

#include "runner_support.hpp"

using namespace sdk;
using namespace sdk::rs;

struct UpdateResultSetup {
  std::shared_ptr<BluefinShieldconexMgmtSDK> client;
  Value data;
  Value idmap;
  Value env;
  bool live = false;
  bool synthetic_only = false;
  long long now = 0;
};

static UpdateResultSetup update_result_basic_setup(const Value& extra) {
  load_env_local();

  std::string entity_data_file = "../.sdk/test/entity/update_result/UpdateResultTestData.json";
  Value entity_data = vs::parse_json(read_file(entity_data_file));

  Value options = vmap({{"entity", getp(entity_data, "existing")}});
  auto client = BluefinShieldconexMgmtSDK::testSDK(options, extra);

  // idmap via transform (upper-cased id name synthetics), matching the donors.
  Value idmap = Struct::transform(
      vlist({Value("update_result01"), Value("update_result02"), Value("update_result03")}),
      vmap({{"`$PACK`", vlist({
        Value(""),
        vmap({
          {"`$KEY`", Value("`$COPY`")},
          {"`$VAL`", vlist({Value("`$FORMAT`"), Value("upper"), Value("`$COPY`")})}
        })
      })}}));
  if (!idmap.is_map()) idmap = vmap();

  Value env = env_override(vmap({
    {"BLUEFINSHIELDCONEXMGMT_TEST_UPDATE_RESULT_ENTID", idmap},
    {"BLUEFINSHIELDCONEXMGMT_TEST_LIVE", Value("FALSE")},
    {"BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN", Value("FALSE")}
  }));

  Value idmap_resolved = Helpers::toMapAny(getp(env, "BLUEFINSHIELDCONEXMGMT_TEST_UPDATE_RESULT_ENTID"));
  if (!idmap_resolved.is_map()) idmap_resolved = idmap;

  bool live = getp(env, "BLUEFINSHIELDCONEXMGMT_TEST_LIVE") == Value("TRUE");

  UpdateResultSetup s;
  s.client = client;
  s.data = entity_data;
  s.idmap = idmap_resolved;
  s.env = env;
  s.live = live;
  s.synthetic_only = false;
  s.now = now_ms();
  return s;
}

static void update_result_entity_instance() {
  auto testsdk = BluefinShieldconexMgmtSDK::testSDK();
  auto ent = testsdk->update_result();
  ASSERT_EQ(ent->getName(), std::string("update_result"), "entity name");
}

static void update_result_entity_stream() {
  // stream() runs the list op through the full pipeline and returns the
  // result items. Seed two entities via test mode; with the streaming feature
  // active it yields the feature's incremental items, else it falls back to
  // the materialised items — either way every item is yielded.
  Value seed = vmap({{"entity", vmap({{"update_result", vmap({
      {"strm01", vmap({{"id", Value("strm01")}})},
      {"strm02", vmap({{"id", Value("strm02")}})}})}})}});
  Value sdkopts = vmap({{"feature",
      vmap({{"streaming", vmap({{"active", Value(true)}})}})}});

  auto strsdk = BluefinShieldconexMgmtSDK::testSDK(seed, sdkopts);
  auto se = strsdk->update_result();
  std::vector<Value> items = se->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)items.size(), 2, "stream yields both seeded items");

  auto plainsdk = BluefinShieldconexMgmtSDK::testSDK(seed, Value::undef());
  auto pe = plainsdk->update_result();
  std::vector<Value> pitems = pe->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)pitems.size(), 2, "fallback stream yields both items");
}

static void update_result_entity_basic() {
  auto setup = update_result_basic_setup(Value::undef());
  std::string mode = setup.live ? "live" : "unit";
  for (const std::string& op : std::vector<std::string>{"create", "list", "update"}) {
    auto sk = is_control_skipped("entityOp", std::string("update_result.") + op, mode);
    if (sk.first) { std::cerr << "skip: " << (sk.second.empty()? "sdk-test-control.json" : sk.second) << "\n"; return; }
  }
  auto client = setup.client;
  // CREATE
  auto update_result_ref01_ent = client->update_result();
  Value update_result_ref01_data = Helpers::toMapAny(getp(Struct::getpath(setup.data, {"new", "update_result"}), "update_result_ref01"));
  if (!update_result_ref01_data.is_map()) update_result_ref01_data = vmap();
  {
    Value update_result_ref01_data_result = update_result_ref01_ent->create(Struct::clone(update_result_ref01_data), Value::undef());
    update_result_ref01_data = Helpers::toMapAny(update_result_ref01_data_result);
    if (!update_result_ref01_data.is_map()) update_result_ref01_data = vmap();
    ASSERT_TRUE(update_result_ref01_data.is_map(), "expected create result to be a map");
    ASSERT_TRUE(!getp(update_result_ref01_data, "id").is_undef(), "expected created entity to have an id");
  }

  // LIST
  Value update_result_ref01_match = vmap();
  Value update_result_ref01_list = update_result_ref01_ent->list(Struct::clone(update_result_ref01_match), Value::undef());
  ASSERT_TRUE(update_result_ref01_list.is_list(), "expected list result to be an array");
  {
    std::vector<Value> found = Struct::select(entity_list_to_data(update_result_ref01_list), vmap({{"id", getp(update_result_ref01_data, "id")}}));
    ASSERT_TRUE(!found.empty(), "expected to find created entity in list");
  }

  // UPDATE
  Value update_result_ref01_data_up0_up = vmap();
  setp(update_result_ref01_data_up0_up, "id", getp(update_result_ref01_data, "id"));
  std::string update_result_ref01_data_up0_markval = std::string("Mark01-update_result_ref01_") + std::to_string(setup.now);
  setp(update_result_ref01_data_up0_up, "billing_id", Value(update_result_ref01_data_up0_markval));
  Value update_result_ref01_resdata_up0_result = update_result_ref01_ent->update(Struct::clone(update_result_ref01_data_up0_up), Value::undef());
  Value update_result_ref01_resdata_up0 = Helpers::toMapAny(update_result_ref01_resdata_up0_result);
  if (!update_result_ref01_resdata_up0.is_map()) update_result_ref01_resdata_up0 = vmap();
  ASSERT_TRUE(update_result_ref01_resdata_up0.is_map(), "expected update result to be a map");
  ASSERT_EQ_VAL(getp(update_result_ref01_resdata_up0, "id"), getp(update_result_ref01_data_up0_up, "id"), "expected update result id to match");
  ASSERT_EQ_VAL(getp(update_result_ref01_resdata_up0, "billing_id"), Value(update_result_ref01_data_up0_markval), "expected billing_id to be updated");

}

int main() {
  T_RUN(update_result_entity_instance);
  T_RUN(update_result_entity_stream);
  T_RUN(update_result_entity_basic);
  return sdktest::summary("update_result_entity_test");
}
